### DISCUSSION TITLE
Is there a way to read and write X86 ports?

### DISCUSSION THREAD

**Discussion by parkertomatoes** [copy/v86#1184] [Q&A] [answered]
Hi all,

I have some old DOS software that communicates through DOS I/O ports ("inp" and "out"). I'd like to emulate this software in V86 and communicate with it from a Javascript function. Is there an exposed API to read and write the emulated V86 ports?

**Comment by chschnell** [comment, marked as answer]
Something like this should work:

```Javascript
YourApplication.Prototype.start_v86 = function()
{
    this.vm = new V86({ ... });

    // Example: Fictional I/O port 0x28
    let port_0x28 = undefined;

    this.vm.bus.register('emulator-started', () => {
        if(port_0x28 === undefined) {      // make sure to register our 0x28 I/O handler only once
            const io = this.vm.v86.cpu.io; // instance of class IO
            port_0x28 = 0;                 // set initial port value

            io.register_read(0x28, this, function() {
                // handle read-access (inp) to port 0x28
                console.log(`inp 0x28: ${port_0x28}`);
                return port_0x28;
            });

            io.register_write(0x28, this, function(out_byte) {
                // handle write-access (out) to port 0x28
                port_0x28 = out_byte;
                console.log(`out 0x28: ${port_0x28}`);
            });
        }
    });
}
```

I just tested this function with guest FreeDOS and QBasic, you will have to see how you weave that into your code.

Note that after creating your V86 instance you have to wait a bit until `this.vm.v86.cpu.io` exists, bus event `emulator-started` which I use here should be safe. Once you have access to the IO object you can register read- and write-access handler for any unused port.

Take a good look at the API of [class IO](https://github.com/copy/v86/blob/master/src/io.js) for the choices you have there, and for usage examples at the source of any emulated hardware subsystem (like [UART](https://github.com/copy/v86/blob/master/src/uart.js) or [PS2](https://github.com/copy/v86/blob/master/src/ps2.js) to name a few), class IO is used all over the place.

**Reply by parkertomatoes** [reply]
Thanks!
